File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    dq = deque([(0, 0, 1)])
    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    visited = [[False for _ in range(m)] for _ in range(n)]
    visited[0][0] = True
    
    while dq:
        cy, cx, step = dq.popleft()
        
        if cy == n - 1 and cx == m - 1:
            return step
        
        for dy, dx in dirs:
            ny = cy + dy
            nx = cx + dx
    
            if 0 <= ny and ny < n and 0 <= nx and nx < m and maps[ny][nx] == 1 and not visited[ny][nx]:
                dq.append((ny, nx, step + 1))
                visited[ny][nx] = True
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    const visited = Array.from({ length: n }, () => Array(m).fill(false));
    const q = [[0, 0, 1]];
    let head = 0;
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]];
    
    visited[0][0] = true;
    
    while (head < q.length) {
        const [cy, cx, step] = q[head];
        head++;
        
        if (cy == n - 1 && cx == m - 1) {
            return step;
        }
        
        for (const dir of dirs) {
            const [dy, dx] = dir;
            const ny = cy + dy;
            const nx = cx + dx;
            
            if (ny >= 0 && ny < n && nx >= 0 && nx < m && !visited[ny][nx] && maps[ny][nx] !== 0) {
                visited[ny][nx] = true;
                q.push([ny, nx, step + 1])
            }
        }
    }
    
    return -1;
}